samcli/hook_packages/terraform/hooks/prepare/resource_linking.py

```diff
@@ -231,9 +231,12 @@ def _link_using_terraform_config(self, source_tf_resource: TFResource, cfn_resou
             )
         if not dest_resources:
             LOG.debug(
-                "There are destination resources defined for for the source resource %s",
+                "There are destination resources defined for for the source resource %s, skipping linking.",
                 source_tf_resource.full_address,
             )
+
+            return
```

Comment:
LGTM in general. I am just wondering why having 2 destinations is throwing an error (at line 229 above) but defining none is skipping this step.

Reply:
[This comment](https://github.com/aws/aws-sam-cli/pull/5196#discussion_r1206143196) explains it in a little more detail, but essentially we don't want to block the execution of some testing cases where the customer may have a partially defined API (eg. defined gateway, but missing some parts of method or integration).

If they were looking to only test their Lambda function using `sam local invoke`, blocking here on missing links would require them to fully define an API Gateway to test their Lambda function.

Reply:
If this method linking layers to functions and also APIGW methods/paths to functions, should we have distinct ways?

Like different part of our logic can call this function (either linking layers to functions or apigw methods/paths to functions), then this method can do whatever it supposed to do. The caller function then can decide whether the case is valid for that purpose or not. I feel like this method is been overloaded with different purposes, and this might lead to a bug in the future. (Of course there is a chance that I mis-understood everything wrong, in that case please ignore my comment :D )

Reply:
This function is a generic function that we used to link different resources. The logic of what to be linked is maintained here https://github.com/aws/aws-sam-cli/blob/develop/samcli/hook_packages/terraform/hooks/prepare/resources/resource_links.py#L29-L81

Reply:
for your comment regarding the exception we raise in line 229, this is related to the Terraform limitations we have, at this point we could not differentiate if the customer is really linking to 2 different resources, or using some conditional expression to choose one of them. As, we do not know what to be done in this case, shall we use the 2 destination resources, or choose one of them, so, we decided to raise an exception to the customer that we do not support this case, and they can resolve this issue by running `terraform apply` command, so we can use a different linking approach where we do not have any issues.

But, for no destination resources found at all, this means that there is no destinations at all to be linked, and should not be an issue (this like if the customer does not have any layers defined for a lambda function), so we do not raise exception here. Sometimes, it can be an issue in the customer terraform project, but we decided not to raise the exception here, and leave this decision to each SAM CLI command to decide if it is an issue or not (Like the customer define a Rest API method, but does not link it to a Lambda function)